readme.md
[![Minimal node version](https://img.shields.io/static/v1?label=node&message=%3E=14.16&logo=node.js&color)](https://nodejs.org/about/releases/)
[![Minimal npm version](https://img.shields.io/static/v1?label=npm&message=%3E=6.14.12&logo=npm&color)](https://github.com/npm/cli/releases)
[![Maintenance](https://img.shields.io/badge/Maintained%3F-yes-green.svg)](https://GitHub.com/stephen-shopopop/node-ts/graphs/commit-activity)
[![Linux](https://svgshare.com/i/Zhy.svg)](https://svgshare.com/i/Zhy.svg)
[![macOS](https://svgshare.com/i/ZjP.svg)](https://svgshare.com/i/ZjP.svg)
[![Visual Studio Code](https://img.shields.io/badge/--007ACC?logo=visual%20studio%20code&logoColor=ffffff)](https://code.visualstudio.com/)

# UUIDv5 generator

## Description

Generate uuidv5 [RFC4122](https://datatracker.ietf.org/doc/html/rfc4122)

```bash
    0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |                          time_low                             |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |       time_mid                |         time_hi_and_version   |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |clk_seq_hi_res |  clk_seq_low  |         node (0-1)            |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |                         node (2-5)                            |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

### Installation

[![Minimal node version](https://img.shields.io/static/v1?label=node&message=%3E=14&logo=node.js&color)](https://nodejs.org/about/releases/)

Add .npmrc file in your repository

> @stephen-shopopop:registry=https://npm.pkg.github.com


```bash
npm install @stephen-shopopop/uuidv5@latest
```

## Usage

```ts
import uuidv5 from "uuidv5";

const data = new TextEncoder().encode("Hello World!");

// Generate a v5 UUID using a namespace and some data.
const myUUID = uuidv5.generate("6ba7b810-9dad-11d1-80b4-00c04fd430c8", data);

// Validate the v5 UUID.
const isValid = uuidv5.validate(myUUID);
```

## UUIDv4

Generate uuidv4 with nodejs > 14 [crypto.randomuuid](https://nodejs.org/docs/latest-v14.x/api/crypto.html#crypto_crypto_randomuuid_options)

```js
const crypto = require('crypto')

// typescript
// import crypto from 'crypto'

const myUUID = crypto.randomUUID()
```

## Contributing

1. npm run build -  Build library.
2. npm run test - Run test with jest.
3. npm run lint - Lint your code.
4. npm run lint:fix - Lint & fix your code.
5. npm run typecheck - Run typescript check.
6. npm run doc - Generate html doc.
7. npm run release - Release library
8. npm run fix - Fix library

### Package maintenance

A modern cli tool that keeps your deps fresh

```bash
npx taze
```



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.1.5](https://github.com/stephen-shopopop/uuidv5/compare/v1.1.4...v1.1.5) (2022-07-31)

### [1.1.4](https://github.com/stephen-shopopop/uuidv5/compare/v1.1.3...v1.1.4) (2022-07-31)


### Bug Fixes

* **bundle:** FIX dependencies ([9a5eda1](https://github.com/stephen-shopopop/uuidv5/commit/9a5eda19459e79fb5a9118559067fc5da0478f3e))

### [1.1.3](https://github.com/stephen-shopopop/uuidv5/compare/v1.1.2...v1.1.3) (2022-07-31)

### [1.1.2](https://github.com/stephen-shopopop/uuidv5/compare/v1.1.1...v1.1.2) (2022-07-31)


### Bug Fixes

* **package:** FIX package json ([0e5647a](https://github.com/stephen-shopopop/uuidv5/commit/0e5647a311c3f2aa440a73cfd9604669397059ab))

### [1.1.1](https://github.com/stephen-shopopop/uuidv5/compare/v1.1.0...v1.1.1) (2022-07-31)

## 1.1.0 (2022-07-31)


### Features

* **uuidv5:** ADD uuidv5 algo ([5715414](https://github.com/stephen-shopopop/uuidv5/commit/57154146be063bc28f1f3a501ed8323580d90d09))


src/index.test.ts
import { describe, expect, test } from '@jest/globals'
import crypto from 'crypto'
import { TextEncoder } from 'util'
import uuidv5 from './index'

describe('[index] uuidv5 generate()', () => {
  test.each`
    namespace | uuid | data | expected
    ${'DNS'} | ${'6ba7b810-9dad-11d1-80b4-00c04fd430c8'} | ${'hello'} | ${'9342d47a-1bab-5709-9869-c840b2eac501'}
    ${'URL'} | ${'6ba7b811-9dad-11d1-80b4-00c04fd430c8'} | ${'hello'} | ${'074171de-bc84-5ea4-b636-1135477620e1'}
    ${'URL'} | ${'6ba7b811-9dad-11d1-80b4-00c04fd430c8'} | ${'http://example.org/page'} | ${'6b19973b-8154-5782-bca0-15e6b730ca00'}
    ${'OID'} | ${'6ba7b812-9dad-11d1-80b4-00c04fd430c8'} | ${'hello'} | ${'4d71d03f-f19b-5d9e-8523-9628ba18063c'}
    ${'X500'} | ${'6ba7b814-9dad-11d1-80b4-00c04fd430c8'} | ${'hello'} | ${'feddb5c6-6b33-5dd7-a4d8-fb98254f583f'}
  `('namespace $namespace returns $expected when $data', ({ namespace, uuid, data, expected }) => {
    expect(
      uuidv5.generate(uuid, new TextEncoder().encode(data))
    ).toEqual(expected)
  })
})

describe('[index] uuidv5 valid()', () => {
  test('9342d47a-1bab-5709-9869-c840b2eac501 is uuid v5 valid', () => {
    expect(uuidv5.validate('9342d47a-1bab-5709-9869-c840b2eac501')).toBeTruthy()
  })

  test('crypto.randomUUID() is uuid v5 not valid', () => {
    expect(uuidv5.validate(crypto.randomUUID())).toBeFalsy()
  })

  test('hello-world is uuid v5 not valid', () => {
    expect(uuidv5.validate('hello-world')).toBeFalsy()
  })
})


src/uuidv5.ts
// Copyright 2021-2022 Shopopop. All rights reserved. ISC license.

import assert from 'assert'
import crypto from 'crypto'

/**
 * Generate a RFC4122 v5 UUID (SHA-1 namespace).
 *
 * ```ts
 * import uuidv5 from "uuidv5";
 *
 * const NAMESPACE_URL = "6ba7b810-9dad-11d1-80b4-00c04fd430c8";
 * const data = new TextEncoder().encode("example.org")
 *
 * const uuid =  uuidv5.generate(NAMESPACE_URL, data);
 * uuid === "886313e1-3b8a-5372-9b90-0c9aee199e5d" // true
 * ```
 *
 * @param namespace The namespace to use, uuid valid
 * @param data The data to hash to calculate the SHA-1 digest for the UUID
 * @returns uuid
 */
export function generate (namespace: string, data: Uint8Array): string {
  // namespace parse hexa
  const space: number[] = []
  namespace.replace(/[a-fA-F0-9]{2}/g, (hex: string): string => {
    space.push(parseInt(hex, 16))
    return ''
  })

  assert(space.length === 16, 'namespace must be a valid UUID')

  // concat Uint8Array namespace and data
  const toHash = new Uint8Array(space.length + data.length)
  toHash.set(space, 0)
  toHash.set(data, space.length)

  const buffer = crypto.createHash('sha1').update(toHash).digest()
  const bytes = new Uint8Array(buffer)

  bytes[6] = (bytes[6] & 0x0f) | 0x50
  bytes[8] = (bytes[8] & 0x3f) | 0x80

  // convert byte to uuid string
  const bits = [...bytes].map(bit => (bit < 0x10 ? '0' : '') + bit.toString(16))

  return [
    ...bits.slice(0, 4),
    '-',
    ...bits.slice(4, 6),
    '-',
    ...bits.slice(6, 8),
    '-',
    ...bits.slice(8, 10),
    '-',
    ...bits.slice(10, 16)
  ].join('')
}

const UUIDV5_REGEX = /^[0-9a-f]{8}-[0-9a-f]{4}-[5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i

/**
 * Validate is an RFC4122 v5 UUID.
 *
 * ```ts
 * import uuidv5 from "uuidv5";
 * import crypto from "crypto"
 *
 * uuidv5.validate(uuidv5.generate("6ba7b810-9dad-11d1-80b4-00c04fd430c8", new Uint8Array())); // true
 * uuidv5.validate(crypto.randomUUID()); // false
 * uuidv5.validate("this-is-not-a-uuid"); // false
 * ```
 */
export function validate (uuid: string): boolean {
  return UUIDV5_REGEX.test(uuid)
}


src/prototype.js
const assert = require('assert')
const crypto = require('crypto')

function generate(namespace, data){
	
	// namespace hexa to decimal
	const space = []
	namespace.replace(/[a-fA-F0-9]{2}/g, (hex) => space.push(parseInt(hex, 16)) && '')

	assert(space.length === 16, "namespace must be a valid UUID")
	
	// concat Uint8Array space and data
	const toHash = new Uint8Array(space.length + data.length)
	toHash.set(space, 0)
	toHash.set(data, space.length)
	
	const buffer = crypto.createHash('sha1').update(toHash).digest()
	const bytes = new Uint8Array(buffer)

    bytes[6] = (bytes[6] & 0x0f) | 0x50
    bytes[8] = (bytes[8] & 0x3f) | 0x80
    
    // bytes to hexa string
    const bits = [...bytes].map(bit => (bit < 0x10 ? '0' : '') + bit.toString(16))
    
    
    return [
      ...bits.slice(0, 4),
      "-",
      ...bits.slice(4, 6),
      "-",
      ...bits.slice(6, 8),
      "-",
      ...bits.slice(8, 10),
      "-",
      ...bits.slice(10, 16),
    ].join("")
}

const UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i

function validate(id) {
  return UUID_RE.test(id);
}

// test
// generate('9999', 'hello')
const c = generate('6ba7b811-9dad-11d1-80b4-00c04fd430c8', 'http://example.org/page')

assert(c === '6b19973b-8154-5782-bca0-15e6b730ca00', 'uuid error')
// console.log(c)


// console.log(globalThis.crypto)
// console.log(crypto.randomUUID()) // nodejs14
